Formula/infer.rb

```diff
@@ -70,39 +60,16 @@ def install
     # Let's try build clang faster
     ENV["JOBS"] = ENV.make_jobs.to_s
 
-    ENV["CLANG_CMAKE_ARGS"] = "-DLLVM_OCAML_INSTALL_PATH=#{`opam var lib`.chomp}/ocaml"
-
     # Release build
     touch ".release"
 
-    # Pin updated dependencies which are required to build on brew ocaml
-    # Remove from this when Infer updates their opam.locked to use at least these versions
-    pinned_deps = {
-      "mlgmpidl"  => "1.2.12",
-      "octavius"  => "1.2.1",
-      "parmap"    => "1.0-rc11",
-      "ppx_tools" => "5.3+4.08.0",
-    }
-    pinned_deps.each { |dep, ver| system "opam", "pin", "add", dep, ver, "--locked" }
-
-    # Unfortunately, opam can't cope if a system ocaml-num happens to be installed.
-    # Instead, we depend on Homebrew's ocaml-num and fool opam into using it.
-    # https://github.com/ocaml/opam-repository/issues/14646
-    system "opam", "pin", "add", "ocamlfind", Formula["ocaml-findlib"].version.to_s, "--locked", "--fake"
-    system "opam", "pin", "add", "num", Formula["ocaml-num"].version.to_s, "--locked", "--fake"
-
-    # Relax the dependency lock on a specific ocaml
-    # Also ignore anything we pinned above
-    ENV["OPAMIGNORECONSTRAINTS"] = "ocaml,ocamlfind,num,#{pinned_deps.keys.join(",")}"
-
-    # Remove ocaml-variants dependency (we won't be using it)
-    inreplace "opam.locked", /^ +"ocaml-variants" \{= ".*?"\}$\n/, ""
-
-    system "opam", "exec", "--", "./build-infer.sh", "all", "--yes", "--user-opam-switch"
-    system "opam", "exec", "--", "make", "install-with-libs"
+    system "./build-infer.sh", "all", "--yes"
+    system "make", "install-with-libs"
```

Comment:
Does this build a custom clang as the file is suggesting? Because that would conflict with https://docs.brew.sh/Acceptable-Formulae#stuff-that-requires-vendored-versions-of-homebrew-formulae

Reply:
No -- Infer *uses* `clang` internally to parse and reason about source code (it is a static analysis tool). The version of Infer is intrinsically tied to a particular revision of `clang`.

This Formula does not build `clang` and then use `clang` to build Infer -- it uses the system compiler to build Infer, and Infer's build process builds a version of `clang` that it then links against.

Does that make sense? Is this acceptable?

Reply:
So, on macOS it would use `LLVM` to build `infer`, and in doing so it would build a `clang` that it infer would then use instead of the system tools?

@Homebrew/maintainers any opinions on this? I know we removed facebook projects for doing these kind of things before.

Reply:
> in doing so it would build a clang that it infer would then use instead of the system tools?
>

Not really ... it builds a `libclang.dylib` that it then links against and uses for obtaining the AST of C/C++ code.

The built `clang` executable is never expected be used by an end user.

Reply:
Just to be clearer:

> The version of Infer is intrinsically tied to a particular revision of clang.
>

It is not expected that Infer would work with the system version of LLVM. It fact, it is more expected that it *won't* work.

Infer has a set of bindings between LLVM and OCaml -- these bindings are hand-written and hand-maintained to meet a specific version of LLVM.

It is not possible to just swap-in another LLVM version and expect these bindings to be compatible.

Reply:
> The built `clang` executable is never expected be used by an end user.

In that case: let's not install it, please.

> It is not expected that Infer would work with the system version of LLVM. It fact, it is more expected that it _won't_ work.

Whether it's expected or not: can you confirm whether it does/does not work with:
- the system Clang
- the Clang from any of the `llvm` or `llvm@*` formulae

If it's tested and known to be definitely broken for all of those: ok. If not: let's try to use one of them.

Reply:
> Not really ... it builds a libclang.dylib that it then links against and uses for obtaining the AST of C/C++ code.
> The built clang executable is never expected be used by an end user.

Doesn't this also mean that if there is no way to "avoid shipping outdated/insecure versions of software that is already a formula"?

Reply:
> In that case: let's not install it, please.

I'm currently building the `@11` JDK build, so don't have an install to check this ... but are you saying just remove the whole `bin` folder from the install? We need the libs, but not the LLVM/`clang` executables.

> If it's tested and known to be definitely broken for all of those: ok. If not: let's try to use one of them.

I 100% agree with the suggestion/intent/etc. with this; I'm just now out of my depth to resolve this, as I don't know how to "trick" Infer to use an out-of-tree build of Clang.

Reply:
> That makes it fundamentally conflict with the goals of a package manager.

I am not sure that's true -- some packages need dependencies at different/older versions; some packages need *patched* dependencies at different/versions.

If Homebrew doesn't allow this (and that's totally fine if you don't!), then it is very possible that Infer is fundamentally incompatible with Homebrew as it does need a *very specific* build of LLVM (and there's a chance it is also patched as well: https://github.com/facebook/infer/tree/master/facebook-clang-plugins/clang/src).

Reply:
> I'm currently building the `@11` JDK build, so don't have an install to check this ... but are you saying just remove the whole `bin` folder from the `clang` install? We need the libs, but not the LLVM/`clang` executables.

If we don't need them: yep. That would reduce my concerns considerably.

> If Homebrew doesn't allow this (and that's totally fine if you don't!), then it is very possible that Infer is fundamentally incompatible with Homebrew as it does need a _very specific_ build of LLVM (and there's a chance it is also patched as well: https://github.com/facebook/infer/tree/master/facebook-clang-plugins/clang/src).

We have found that multiple Facebook projects in the past are designed in such a way to make them very hard to package securely and reproducibly (and the upstream projects have been less than receptive in helping us).

Reply:
> If we don't need them: yep. That would reduce my concerns considerably.

Should be fine; I'll try this once my `@11` JDK build has finished (which takes about ~2 hours per build 🤦)

Reply:
Thanks very much for your work on this, @andrewvaughanj. When you get around to it, this line can be removed:

https://github.com/Homebrew/homebrew-core/blob/ed18465cacb1eb1d46df543ae7f0c59a1cdae71b/Formula/infer.rb#L22

(I do agree with the comments above re the superfluous `clang` executable and trying to see if we can link with an existing `libclang`, etc, though)

Reply:
> When you get around to it, this line can be removed:

Already gone in my local, unpushed version 👼

Reply:
Thanks for this interesting discussion!

I'm afraid deleting the clang binaries won't work: Infer needs them as the C/ObjC/C++ frontends are all based on running clang+the infer plugin to extract the AST of source files. This is how clang plugins in general work: they need clang to run together with the plugin. The version of clang used by infer in its current version is stock clang 11 + a handful of patches. Without these patches the Infer clang plugin doesn't work either as clang normally tries to do much more than just parsing even when we only need the AST. For this reason we cannot make infer build against the packaged llvm either, unfortunately.

I'd like to reiterate that the modified clang that Infer uses under the hood is well-hidden from users of Infer and isolated from the rest of system so it doesn't affect anything else. We regularly update the version of clang too so I don't think there are security concerns (I'm not sure I understand the reproducibility concerns, but perhaps that wasn't directed at this PR specifically).

Reply:
> The version of clang used by infer in its current version is stock clang 11 + a handful of patches.

Are there any efforts to get these patches upstreamed?



> I'd like to reiterate that the modified clang that Infer uses under the hood is well-hidden from users of Infer and isolated from the rest of system so it doesn't affect anything else.

Can you elaborate on how it's well-hidden/isolated?

Reply:
> Can you elaborate on how it's well-hidden/isolated?

The live in a directory such as `/usr/local/Cellar/infer/1.1.0/lib/infer/facebook-clang-plugins/clang/install/bin` (which isn't placed on `$PATH`) -- they *are not* put into `/usr/local/bin`.

Reply:
> Are there any efforts to get these patches upstreamed?

That's a good question. In their current form some of these patches are not mergeable as-is as they disable some clang functionalities(!) but I imagine we could find a common ground with clang's needs, perhaps via a command-line option. We are not clang devs but we'll try and find help to hopefully submit something upstream. Thanks for the suggestion!

Reply:
Please do find a way to get these patches upstreamed in some form. Homebrew doesn't like third-party patches (they tend to break things in unexpected and sometimes very damaging ways), and it would be good for users to not have to have several `clang` installations on their system just to run the software they use.